gnovm/stdlibs/std/emit_event.go

```diff
@@ -0,0 +1,75 @@
+package std
+
+// ref: https://github.com/gnolang/gno/pull/575
+// ref: https://github.com/gnolang/gno/pull/1833
+
+import (
+	"encoding/json"
+	"errors"
+
+	gno "github.com/gnolang/gno/gnovm/pkg/gnolang"
+	abci "github.com/gnolang/gno/tm2/pkg/bft/abci/types"
+)
+
+func X_emit(m *gno.Machine, typ string, attrs []string) {
+	eventAttrs, err := attrKeysAndValues(attrs)
+	if err != nil {
+		m.Panic(typedString(err.Error()))
+	}
+
+	pkgPath := CurrentRealmPath(m)
+	fnIdent := getPrevFunctionNameFromTarget(m, "Emit")
+
+	evt := NewGnoEvent(typ, pkgPath, fnIdent, eventAttrs...)
+	ctx := m.Context.(ExecContext)
+	ctx.EventLogger.EmitEvent(evt)
+
+	str, err := json.Marshal(evt)
+	if err != nil {
+		m.Panic(typedString(err.Error()))
+	}
+
+	ctx.EmittedEvents = abci.EventString(str)
+}
+
+func attrKeysAndValues(attrs []string) ([]gnoEventAttribute, error) {
+	attrLen := len(attrs)
+	if attrLen%2 != 0 {
+		return nil, errors.New("cannot pair attributes due to odd count")
```

Comment:
Nitpick, you can extract this error into a package-level err, ex:

`var errInvalidGnoEventAttrs = errors.New("cannot pair attributes due to odd count")`

Reply:
changed